src/batch/src/execution/grpc_exchange.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{Debug, Formatter};
+
+use futures::StreamExt;
+use risingwave_common::array::DataChunk;
+use risingwave_common::error::{Result, ToRwResult};
+use risingwave_pb::batch_plan::{ExchangeSource as ProstExchangeSource, TaskOutputId};
+use risingwave_pb::task_service::{ExecuteRequest, GetDataResponse};
+use risingwave_rpc_client::{ComputeClient, ExchangeSource};
+use tonic::Streaming;
+
+use crate::task::{BatchTaskContext, BatchTaskContextType};
+
+/// Use grpc client as the source.
+pub struct GrpcExchangeSource<C: BatchTaskContext> {
+    stream: Streaming<GetDataResponse>,
+
+    task_output_id: TaskOutputId,
+    context: C,
+}
+
+impl<C: BatchTaskContext> GrpcExchangeSource<C> {
+    pub async fn create(exchange_source: ProstExchangeSource, context: C) -> Result<Self> {
+        let addr = exchange_source.get_host()?.into();
+        let task_output_id = exchange_source.get_task_output_id()?.clone();
+        let task_id = task_output_id.get_task_id()?.clone();
+        let client = ComputeClient::new(addr).await?;
+        let stream = match context.context_type() {
+            // When in the local execution mode, `GrpcExchangeSource` would send out
+            // `ExecuteRequest` and get the data chunks back in a single RPC.
+            BatchTaskContextType::Frontend => {
+                let plan = exchange_source.plan.clone();
+                let execute_request = ExecuteRequest {
+                    task_id: Some(task_id),
+                    plan,
+                    epoch: exchange_source.epoch,
+                };
+                client.execute(execute_request).await?
+            }
+            BatchTaskContextType::ComputeNode => client.get_data(task_output_id.clone()).await?,
```

Comment:
I think we can remove the `BatchTaskContextType` if we have `one_of` in plan proto.